### PR TITLE
Integration test for creating volumes in superset

### PR DESCRIPTION
**Pull request by aelkiss**
Addresses #2 

- Tests that volumes are created on the filesystem and contain only the expected files.
- Refactors dataset_cli to use `Dataset::HathiTrust::Configuration`
- Refactors/simplifies `Dataset::HathiTrust::Configuration` to only do what's currently needed

Future work (will open issues)
- Test that no unexpected volumes are created on the filesystem (to complete #2)
- datasets_cli is really just an integration test wrapper and needs to at least be able to specify a config file and inject a "last run date" before it's complete
-  we should also test logging/reporting in the integration test, but we need to merge #49 before we can test that

### CODE REVIEW

File: config/config.example.yml [hunk 1]
```diff
@@ -3,21 +3,16 @@
 # This is the config.yml to use when using the
 # Datasets::Hathitrust::Configuration.
```
Comment (grosscol) [inline config/config.example.yml]: I would be in favor of removing this file altogether as we already have the integration test config.  The example config for other projects has been functional for the external components in an out of the box fashion where you could copy config.yml.example to config.yml and it works in the demo case in the readme.

File: config/config.example.yml [hunk 2]
```diff
@@ -3,21 +3,16 @@
...
-    adapter: mysql2
-    host: blah
-    etc: etc
-  feed:
-    adapter: mysql
-    host: blah
-    etc: etc
-
+  adapter: mysql2
+  host: blah
```
Comment (grosscol) [inline config/config.example.yml]: If we keep this, let's use an example string that could conceiveably work like localhost:3306

File: config/config.example.yml [hunk 3]
```diff
@@ -3,21 +3,16 @@
...
+  etc: etc
```
Comment (grosscol) [inline config/config.example.yml]: Is there an example of a value here that would actually be used?

File: config/hathitrust_config.rb [hunk 4]
```diff
@@ -11,33 +11,26 @@ def volume_writer
...
+          PairtreePathResolver.new(Pathname.new(src_parent_dir))
       end
 
       def volume_repo
         @volume_repo ||=
           subsets.map do |subset|
             [subset, subset_volume_repo]
           end
-          .concat([superset, superset_volume_repo])
+          .concat([[superset, superset_volume_repo]])
```
Comment (grosscol) [inline config/hathitrust_config.rb]: This is basically appending [:full, "Rights+Feed"] to the array of arrays.  When I replace the volume repos with strings, the result looks like this:
```
superset
=> :full
superset_volume_repo
=> "Rights+Feed"
subset_volume_repo
=> "Rights"
volume_repo             
=> [[:pd, "Rights"], [:pd_open, "Rights"], [:pd_world, "Rights"], [:pd_world_open, "Rights"]]
volume_repo.concat([[superset, superset_volume_repo]])
=> [[:pd, "Rights"], [:pd_open, "Rights"], [:pd_world, "Rights"], [:pd_world_open, "Rights"], [:full, "Rights+Feed"]]
```
Seeing that all makes sense now.  I don't have a good handle on what about this is irking me, but something here blocked me from grok'ing it immediately.  Seems like the [set, set_volume_repo] are inexorably linked. 

Perhaps it's a naming issue.  I think of "volume repo" as a repository of volumes.  This is actually a mapping of the volume subset to the rights database?  or in the case of the full set, the rights and rights and feed backend database? 

 Refactor later I guess.

File: config/hathitrust_config.rb [hunk 5]
```diff
@@ -50,20 +43,24 @@ def filter
...
+      end
+
       private
 
       def subset_volume_repo
-        @rights_volume_repo ||= RightsVolumeRepo.new(rights_db_connection)
+        @rights_volume_repo ||= Repository::RightsVolumeRepo.new(db_connection)
       end
 
       def superset_volume_repo
```
Comment (grosscol) [inline config/hathitrust_config.rb]: The naming of this and subset_volume_repo caused me a bit of confusion, but the functionality of it appear sound.  Refactor this naming later.

File: lib/datasets_cli.rb [hunk 6]
```diff
@@ -17,42 +19,24 @@ class CLI < Thor
     # Tasks
     desc "all", "Do all the dataset operations."
     def all
-      configure unless configured?
+      # TODO: take config yml as a cmdline option
+      config = Datasets::HathiTrust::Configuration.from_yaml(File.join(APP_ROOT,"config/integration.yml"))
```
Comment (grosscol) [inline lib/datasets_cli.rb]: Going this route, we can drop the config gem.

File: spec/datasets_cli_spec.rb [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+      pairtree_prefix = File.join(repo_path,"obj/test/pairtree_root/00")
+
+      old_vol_file = File.join(pairtree_prefix,"1","001","001.zip")
+      FileUtils.touch(old_vol_file,mtime: OLD_TIMESTAMP)
+
+      new_vol_file = File.join(pairtree_prefix,"2","002","002.zip")
+      FileUtils.touch(new_vol_file,mtime: NEW_TIMESTAMP)
+    end
+
+    def add_test_data(db)
```
Comment (grosscol) [inline spec/datasets_cli_spec.rb]: This is fine and works.  My opinion is that this db should be a fixture or factory.  Given the simplicity and static nature of the project, I prefer fixture.

Comment (aelkiss) [inline spec/datasets_cli_spec.rb]: Right, the static config is a fixture (schema and rights attributes) although there's some functionality for it to be done as a factory too. The resetting every time here is a little bit of a belt and suspenders approach to making sure things haven't changed while still giving flexibility to monkey with some of the timestamps in the db on a per-test basis.

File: lib/datasets_cli.rb (hunk 6 above)
Comment (aelkiss) [inline lib/datasets_cli.rb]: Will remove

File: config/hathitrust_config.rb (hunk 4 above)
Comment (aelkiss) [inline config/hathitrust_config.rb]: Yeah, this is a bit confusing and I already mentioned it to @malakai97. Added an issue #55

File: config/config.example.yml (hunk 1 above)
Comment (aelkiss) [inline config/config.example.yml]: I'll just move integration.yml over top config.example.yml.

Comment (aelkiss) [inline config/config.example.yml]: Done

**Comment by aelkiss** [comment]
What advantages do you see to mocking the filesystem other than speeding up the integration tests? The integration tests are definitely the slowest part right now, but the whole suite still runs in 2.36 seconds for me... In any case it seems essential to me to actually hit the filesystem in at least one test since interaction with disk is a central aspect of this project. It already uncovered issues like the `mtime` vs. `ctime` thing where if we mocked it we might have mocked it incorrectly.

**Comment by grosscol** [comment]
My implicit cognition was speed an not testing that FileUtils or File was working.  The ctime vs mtime thing is a good point.  _I guess we can wait the *whole* 2.3 seconds for the test to run_

**Comment by aelkiss** [comment]
Right, not important to test that File or FileUtils is working correctly as such, but it does seem important in an integration test to make sure that we're using all the components together correctly and that the whole end-to-end process work as expected.

File: lib/datasets/configuration.rb [hunk 8]
```diff
@@ -11,7 +11,7 @@ def initialize(hash = {})
     end
 
     def self.from_yaml(path)
-      new(YAML.load(path))
+      new(YAML.load(File.read(path)))
```
Comment (malakai97) [inline lib/datasets/configuration.rb]: Why?

File: lib/datasets/volume.rb [hunk 9]
```diff
@@ -32,5 +32,9 @@ def to_s
       "#{namespace}.#{id} #{right} #{access_profile}"
     end
 
+    def hash
```
Comment (malakai97) [inline lib/datasets/volume.rb]: What is this for?  It's not used, and `#eql?` is already defined.

Comment (aelkiss) [inline lib/datasets/volume.rb]: Needs to be defined for `Set` to work properly. Might be a comment in the spec for this?

Comment (aelkiss) [inline lib/datasets/volume.rb]: ah, no, only explained in the commit for it (https://github.com/hathitrust/datasets/pull/51/commits/dafdd935ad521b78714f84060effc28200089216). I could add a comment in `Volume`

Comment (malakai97) [inline lib/datasets/volume.rb]: Oh right.  The to_s threw me off.

**Pull request merged**